src/dmd/dmodule.d

```diff
@@ -536,6 +537,10 @@ extern (C++) final class Module : Package
         }
         m = m.parse();
 
+        // verify the module name matches the imported module name
+        if (!packages.equals((m.md is null) ? null : m.md.packages))
+            m.error(loc, "from file %s must be imported with 'import %s;'", m.srcfile, m.toPrettyChars());
```

Comment:
Sadly, I think this will introduce a regression, but probably one D needs.

 - This needs to specified in the specification, if it's not already.  
 - This probably needs to go through the deprecation process.  That stinks, but it probably needs to be done.
 - I think this will require a changelog entry

That's just my opinion/perception.  I'd be happy to fast-track this as I think it's the right thing to do to enforce reliable usage of the language, but that's not my call.